args = commandArgs(trailingOnly = TRUE)

library(numbat)
library(data.table)
library(Matrix)

count_mat = as(readMM(args[1]),"CsparseMatrix")#as.matrix(fread(args[1],skip=2)) #readRDS(args[1])
features = fread(args[5],header = F)
# resolve duplicated gene names
features$V2[features$V1=="ENSG00000285053"] = "GGPS1-TBCE" #"TBCE"
features$V2[features$V1=="ENSG00000261186"] = "LINC01238.2" #two genes with same alias, ambiguous "LINC01238"
features$V2[features$V1=="ENSG00000271858"] = "CACNA2D2-anti" #"CYB561D2"
features$V2[features$V1=="ENSG00000280987"] = "MATR3.2" #"MATR3" two genes         
features$V2[features$V1=="ENSG00000234229"] = "LINC01505.2"  #"LINC01505"
features$V2[features$V1=="ENSG00000284024"] = "HSPA14.2" #"HSPA14"
features$V2[features$V1=="ENSG00000261480"] = "GOLGA8M.2" #"GOLGA8M"
features$V2[features$V1=="ENSG00000286070"] = "GGT1.2" #"GGT1"          
features$V2[features$V1=="ENSG00000271147"] = "ARMCX5-GPRASP2.2" #"ARMCX5-GPRASP2" 
features$V2[features$V1=="ENSG00000269226"] = "TMSB15B.2" #"TMSB15B"

barcodes = fread(args[6],header = F)
colnames(count_mat) = barcodes$V1
rownames(count_mat) = features$V2

df_allele = fread(args[2],h=T) #readRDS(args[2])

out_dir = args[3]
ncores = as.numeric(args[4])

# print arguments
head(count_mat)
head(df_allele)
head(ref_hca)

print(ncores)
print(out_dir)

# check data
dim(count_mat)
  
# 
out = run_numbat(
  count_mat, # gene x cell integer UMI count matrix 
  ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
  df_allele, # allele dataframe generated by pileup_and_phase script
  genome = "hg38",
  t = 1e-5,
  ncores = ncores,
  plot = TRUE,
  out_dir = out_dir
)

